src/librustc_mir/borrow_check.rs

```diff
@@ -373,10 +373,41 @@ impl<'cx, 'gcx, 'tcx> DataflowResultsConsumer<'cx, 'tcx> for MirBorrowckCtxt<'cx
                                      Consume, (value, span), flow_state);
             }
 
-            TerminatorKind::Goto { target: _ } |
             TerminatorKind::Resume |
             TerminatorKind::Return |
-            TerminatorKind::GeneratorDrop |
+            TerminatorKind::GeneratorDrop => {
+                // Returning from the function implicitly kills storage for all locals and statics.
+                // Often, the storage will already have been killed by an explicit
+                // StorageDead, but we don't always emit those (notably on unwind paths),
+                // so this "extra check" serves as a kind of backup.
+                let domain = flow_state.borrows.base_results.operator();
+                for borrow in domain.borrows() {
+                    let root_lvalue = self.prefixes(
+                        &borrow.lvalue,
+                        PrefixSet::All
+                    ).last().unwrap();
+                    match root_lvalue {
+                        Lvalue::Static(_) => {
+                            self.access_lvalue(
+                                ContextKind::StorageDead.new(loc),
+                                (&root_lvalue, self.mir.source_info(borrow.location).span),
+                                (Deep, Write(WriteKind::StorageDeadOrDrop)),
```

Comment:
The other thing I'm not sure about is whether this should be a shallow write instead.

Reply:
This should be a drop - basically a shallow write for structs without a destructor, and a deep write for structs without one.

Reply:
Ah so this code only runs for thread-locals. In that case, I suppose there won't be much harm from it remaining a deep write.